File: Solutions/206. Reverse Linked List/206. Reverse Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 //iterative
func reverseList(head *ListNode) *ListNode {
    var newhead *ListNode=nil
    for head!=nil{
        next:=head.Next
        head.Next=newhead
        newhead=head
        head=next
    }
    return newhead
}


//recursive
func reverseList(head *ListNode) *ListNode {
    return reverList(head,nil)
}
func reverList(head,newhead *ListNode) *ListNode{
    if head==nil{
        return newhead
    }
    next:=head.Next
    head.Next=newhead
    return reverList(next,head)
}

File: Solutions/206. Reverse Linked List/206. Reverse Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

//recursive
struct ListNode* foo(struct ListNode*,struct ListNode*);

struct ListNode* reverseList(struct ListNode* head){
    return foo(head,NULL);
}
struct ListNode* foo(struct ListNode* head,struct ListNode* newhead){
    if(!head){
        return newhead;
    }
    struct ListNode* tmp=head->next;
    head->next=newhead;
    return foo(tmp,head);
}

//iterative
struct ListNode* reverseList(struct ListNode* head){
    struct ListNode* newhead=NULL;
    while(head){
        struct ListNode* tmp=head->next;
        head->next=newhead;
        newhead=head;
        head=tmp;
    }
    return newhead;
}